app/models/tag.rb

```diff
@@ -243,6 +243,18 @@ def update_synonym_associations
     end
   end
 
+  def all_category_ids
+    @all_category_ids ||=
+      categories.pluck(:id) +
+        tag_groups.includes(:categories).map { |tg| tg.categories.map(&:id) }.flatten
```

Comment:
You can do `map` and `flatten` in one "operation" using `flat_map` 👍 

```suggestion
    @all_category_ids ||=
      categories.pluck(:id) +
        tag_groups.includes(:categories).flat_map { |tg| tg.categories.map(&:id) }
```